File: staff-engagement-frontend/src/app/tasks/task.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

import { environment } from '../../environments/environment';
import {
  CreateStandaloneTaskRequest,
  CreateTaskFromInteractionRequest,
  Task,
  UpdateTaskRequest,
  UpdateTaskStatusRequest
} from './task.models';

@Injectable({ providedIn: 'root' })
export class TaskService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = `${environment.apiUrl}/tasks`;

  createFromInteraction(request: CreateTaskFromInteractionRequest): Observable<Task> {
    return this.http.post<Task>(`${this.baseUrl}/from-interaction`, request);
  }

  createStandalone(request: CreateStandaloneTaskRequest): Observable<Task> {
    return this.http.post<Task>(this.baseUrl, request);
  }

  getMyTasks(): Observable<Task[]> {
    return this.http.get<Task[]>(`${this.baseUrl}/mine`);
  }

  getByEmployee(employeeId: number): Observable<Task[]> {
    return this.http.get<Task[]>(this.baseUrl, { params: { relatesToId: employeeId } });
  }

  updateStatus(id: number, request: UpdateTaskStatusRequest): Observable<Task> {
    return this.http.patch<Task>(`${this.baseUrl}/${id}/status`, request);
  }

  update(id: number, request: UpdateTaskRequest): Observable<Task> {
    return this.http.put<Task>(`${this.baseUrl}/${id}`, request);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }
}

File: staff-engagement-backend/src/main/java/com/psybergate/staff_engagement/task/TaskService.java
package com.psybergate.staff_engagement.task;

import com.psybergate.staff_engagement.employee.Employee;
import com.psybergate.staff_engagement.task.dto.CreateStandaloneTaskRequest;
import com.psybergate.staff_engagement.task.dto.CreateTaskFromInteractionRequest;
import com.psybergate.staff_engagement.task.dto.TaskResponse;
import com.psybergate.staff_engagement.task.dto.UpdateTaskRequest;
import com.psybergate.staff_engagement.task.dto.UpdateTaskStatusRequest;
import java.util.List;

/** Module boundary for the task domain. */
public interface TaskService {

    TaskResponse createFromInteraction(CreateTaskFromInteractionRequest request, Employee currentEmployee);

    TaskResponse createStandalone(CreateStandaloneTaskRequest request, Employee currentEmployee);

    List<TaskResponse> getMyTasks(Employee currentEmployee);

    List<TaskResponse> getByRelatesTo(Long employeeId);

    List<TaskResponse> getAllTasks();

    TaskResponse updateStatus(Long id, UpdateTaskStatusRequest request, Employee currentEmployee);

    TaskResponse update(Long id, UpdateTaskRequest request, Employee currentEmployee);

    void delete(Long id, Employee currentEmployee);
}
